### ISSUE TITLE
Need support for SecurityGroupRule

### ISSUE THREAD

**Issue by rbranche** [crossplane-contrib/provider-aws#1316]
### What resource do you need?
`SecurityGroupRule`

https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_SecurityGroupRule.html

### What is your use case?
There doesn't appear to be a way to specify a SecurityGroup which has egress/ingress rules that reference itself.

With the following `Composition.spec.resources.base[0]`:

```
  - base: 
       apiVersion: ec2.aws.crossplane.io/v1beta1
        kind: SecurityGroup
        metadata: {}
        spec:
          forProvider:
            description: placeholder
            egress:
              - ipProtocol: tcp
                fromPort: 6380
                toPort: 6380
                userIdGroupPairs:
                - description: Allowing Cache port Outbound Access
                  groupIdSelector:
                    matchControllerRef: true
            groupName: placeholder
            ingress:
              - ipProtocol: tcp
                fromPort: 6380
                toPort: 6380
                userIdGroupPairs:
                - description: Allowing Cache port Inbound Access
                  groupIdSelector:
                    matchControllerRef: true
            region: placeholder
            tags: []
          providerConfigRef:
            name: placeholder
          writeConnectionSecretToRef:
            name: placeholder
            namespace: placeholder
          deletionPolicy: Delete
```

I'm getting an error where the `groupId` is being filled out with the `groupName` instead of the `groupId`:

```
  Warning  CannotObserveExternalResource  9s (x11 over 51s)  managed/securitygroup.ec2.aws.crossplane.io  failed to describe SecurityGroup: api error InvalidGroupId.Malformed: Invalid id: "crossplane-poc-rediscluster" (expecting "sg-...")
```

Full describe:
```
lekonia:aws-provider{(ff21156...)} branche$ kubectl describe securitygroup crossplane-poc-7ccrk-ns5zx
Name:         crossplane-poc-7ccrk-ns5zx
Namespace:
Labels:       crossplane.io/claim-name=crossplane-poc
              crossplane.io/claim-namespace=branche-crossplane-poc
              crossplane.io/composite=crossplane-poc-7ccrk
Annotations:  crossplane.io/external-name: crossplane-poc-rediscluster
API Version:  ec2.aws.crossplane.io/v1beta1
Kind:         SecurityGroup
Metadata:
  Creation Timestamp:  2022-05-20T22:09:43Z
  Generate Name:       crossplane-poc-7ccrk-
  Generation:          12
  Managed Fields:
    API Version:  ec2.aws.crossplane.io/v1beta1
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .:
          f:crossplane.io/external-name:
        f:generateName:
        f:labels:
          .:
          f:crossplane.io/claim-name:
          f:crossplane.io/claim-namespace:
          f:crossplane.io/composite:
        f:ownerReferences:
          .:
          k:{"uid":"e83e4115-c3a4-4ee3-a4c5-bc76eafee546"}:
            .:
            f:apiVersion:
            f:controller:
            f:kind:
            f:name:
            f:uid:
      f:spec:
        .:
        f:deletionPolicy:
        f:forProvider:
          .:
          f:description:
          f:groupName:
          f:region:
        f:providerConfigRef:
          .:
          f:name:
        f:writeConnectionSecretToRef:
          .:
          f:name:
          f:namespace:
    Manager:      crossplane
    Operation:    Update
    Time:         2022-05-20T22:10:25Z
    API Version:  ec2.aws.crossplane.io/v1beta1
    Fields Type:  FieldsV1
    fieldsV1:
      f:spec:
        f:forProvider:
          f:egress:
          f:ingress:
      f:status:
        .:
        f:atProvider:
          .:
          f:ownerId:
          f:securityGroupID:
        f:conditions:
    Manager:    crossplane-aws-provider
    Operation:  Update
    Time:       2022-05-20T22:10:25Z
  Owner References:
    API Version:     awsblueprints.io/v1alpha1
    Controller:      true
    Kind:            XRedisCluster2
    Name:            crossplane-poc-7ccrk
    UID:             e83e4115-c3a4-4ee3-a4c5-bc76eafee546
  Resource Version:  451258114
  UID:               f3d15440-7580-45ef-9e30-8c5efb13676e
Spec:
  Deletion Policy:  Delete
  For Provider:
    Description:  SecurityGroup for crossplane-poc RedisCluster
    Egress:
      From Port:    6380
      Ip Protocol:  tcp
      To Port:      6380
      User Id Group Pairs:
        Description:  Allowing Cache port Outbound Access
        Group Id:     crossplane-poc-rediscluster
        Group Id Ref:
          Name:  crossplane-poc-7ccrk-ns5zx
        Group Id Selector:
          Match Controller Ref:  true
    Group Name:                  crossplane-poc-rediscluster
    Ingress:
      From Port:    6380
      Ip Protocol:  tcp
      To Port:      6380
      User Id Group Pairs:
        Description:  Allowing Cache port Inbound Access
        Group Id:     crossplane-poc-rediscluster
        Group Id Ref:
          Name:  crossplane-poc-7ccrk-ns5zx
        Group Id Selector:
          Match Controller Ref:  true
    Region:                      us-east-1
  Provider Config Ref:
    Name:  aws-provider
  Write Connection Secret To Ref:
    Name:       crossplane-poc-rediscluster-sg
    Namespace:  branche-crossplane-poc
Status:
  At Provider:
    Owner Id:
    Security Group ID:
  Conditions:
    Last Transition Time:  2022-05-20T22:09:46Z
    Message:               observe failed: failed to describe SecurityGroup: api error InvalidGroupId.Malformed: Invalid id: "crossplane-poc-rediscluster" (expecting "sg-...")
    Reason:                ReconcileError
    Status:                False
    Type:                  Synced
Events:
  Type     Reason                         Age                From                                         Message
  ----     ------                         ----               ----                                         -------
  Warning  CannotObserveExternalResource  9s (x11 over 51s)  managed/securitygroup.ec2.aws.crossplane.io  failed to describe SecurityGroup: api error InvalidGroupId.Malformed: Invalid id: "crossplane-poc-rediscluster" (expecting "sg-...")
```

This appears to be because the `groupId` is not fully filled out in the resource before calling to the API (and correctly, because it can't be from what I can tell).

-----

In our CloudFormation templates, it appears that we are creating a SecurityGroup separate from the SecurityGroupRules:

```
    "Resources": {
...
        "MicroSgsAckSystemRdsRule4fca11743abee7038e08e969a481b39c": {
            "Properties": {
                "Description": "proteus-aws-operator Ingress",
                "FromPort": 3306,
                "GroupId": {
                    "Ref": "MicroSgsGroupAckSystemRds"
                },
                "IpProtocol": "tcp",
                "SourceSecurityGroupId": {
                    "Ref": "MicroSgsGroupAckSystemRds"
                },
                "ToPort": 3306
            },
            "Type": "AWS::EC2::SecurityGroupIngress"
        },
        "MicroSgsAckSystemRdsRulef85d29f00c6348f44a71be2b2f583ed8": {
            "Properties": {
                "Description": "proteus-aws-operator Egress",
                "DestinationSecurityGroupId": {
                    "Ref": "MicroSgsGroupAckSystemRds"
                },
                "FromPort": 3306,
                "GroupId": {
                    "Ref": "MicroSgsGroupAckSystemRds"
                },
                "IpProtocol": "tcp",
                "ToPort": 3306
            },
            "Type": "AWS::EC2::SecurityGroupEgress"
        },
...
        "MicroSgsGroupAckSystemRds": {
            "Properties": {
                "GroupDescription": "proteus-aws-operator access for RDS",
                "SecurityGroupEgress": [],
                "SecurityGroupIngress": [],
                "Tags": [
                    {
                        "Key": "Name",
                        "Value": "micro-preview-useast1-sgs-ack-system-rds"
                    }
                ],
                "VpcId": {
                    "Ref": "VpcId"
                }
            },
            "Type": "AWS::EC2::SecurityGroup"
        },
```

### Would you be willing to contribute it using [code generator](https://github.com/crossplane/provider-aws/blob/master/CODE_GENERATION.md)?

I would likely need some help to get the `groupIdSelector.matchControllerRef` working properly on the SecurityGroupRule code generation.

**Comment by chlunde** [comment]
FYI there's a PR which will support self-referencing groups without SecurityGroupRule objects:
https://github.com/crossplane/crossplane-runtime/pull/328

**Comment by nabuskey** [comment]
I can confirm self referencing SG rules works after the pull request mentioned above. 
```
- description: test
  groupIdSelector:
    matchControllerRef: true
    policy:
      resolution: Optional
      resolve: Always
```

**Comment by haarchri** [comment]
Possible since #1361 was merged - will be included in v0.29.0 release
